client/src/components/uikit/TextInput/index.tsx

```diff
@@ -1,266 +1,114 @@
-      className,
-      style,
-      as,
-      ...rest
-    }: Props,
-    ref: any
-  ) => {
-    const localRef = useRef<HTMLInputElement>(null);
-    const [secureTextVisible, setSecureTextVisible] = useState(false);
-    const [isFocused, setFocused] = useState(false);
-    const onFocus = () => setFocused(true);
-    const onBlur = () => setFocused(false);
-
-    const { labelProps, inputProps } = useTextField(
-      {
-        ...(rest as any),
-        label,
-        placeholder,
-        value,
-        isRequired,
-        onFocus,
-        onBlur,
-        type: allowSecureTextToggle ? secureTextVisible ? 'text' : 'password' : type, // prettier-ignore
-      },
-      localRef
-    );
-
-    const mergedProps = mergeProps(inputProps, rest);
+/**
+ * Ref: https://react-spectrum.adobe.com/react-aria/TextField.html
+ */
+const TextInput = forwardRef<HTMLInputElement, Props>(
+  ({ label, icon, description, errorMessage, placeholder, ...rest }, ref) => {
+    const [passwordVisible, setPasswordVisible] = useState(false);
+    const isPassword = rest.type === 'password';
 
     return (
-      <Wrapper className={className} style={style}>
-        <Label
-          {...labelProps}
-          isValid={isValid}
-          isAnimated={isFocused || !!value}
-          hasIcon={!!icon}
-        >
-          {label}
-          {isRequired && <span aria-hidden="true">*</span>}
-        </Label>
+      <Wrapper {...rest} validationState={errorMessage ? 'invalid' : 'valid'}>
+        <Label data-required={rest.isRequired}>{label}</Label>
 
-        <InputWrapper isValid={isValid}>
-          {icon && (
-            <IconWrapper>
-              <Icon icon={icon} size={20} color="muted1" />
-            </IconWrapper>
-          )}
+        <InputWrapper>
+          {icon && <InputIconLeft icon={icon} size={20} color="muted1" />}
 
           <Input
-            {...mergedProps}
-            as={as}
-            ref={mergeRefs([localRef, ref])}
-            onChange={onChange}
-            hasIcon={!!icon}
+            ref={ref}
+            placeholder={placeholder}
+            data-password={isPassword || undefined}
+            type={
+              isPassword ? (passwordVisible ? 'text' : 'password') : rest.type
+            }
           />
 
-          {allowSecureTextToggle && (
-            <FocusRing focusRingClass="secure-text-button-focus">
-              <SecureTextButton
-                type="button"
-                onClick={() => setSecureTextVisible(p => !p)}
-              >
-                <Icon
-                  icon={secureTextVisible ? FiEyeOff : FiEye}
-                  size={20}
-                  color="muted1"
-                />
-              </SecureTextButton>
-            </FocusRing>
+          {isPassword && (
+            <PasswordToggleButton
+              isSelected={passwordVisible}
+              onChange={setPasswordVisible}
+              aria-label="Show password"
```

Comment:
Translate with `t` Lingui helper here. Note that it requires you to also add `useLingui()` in the component so that it works correctly with macros.

Reply:
What do you mean with using the hook, IIRC we don't need to pass it to `t` or anything, and for example `/home/jones/work/full-stack-template/client/src/components/common/ColorModeButton.tsx` doesn't use the hook but uses the macro?

Reply:
I might have forgotten to add `useLingui` in some places 😄

Version 4 of Lingui added this requirement when using `t`: https://lingui.dev/tutorials/react-patterns#element-attributes-and-string-only-translations

Reply:
Ah that's good to know, I think it wasn't required in `lingui@v3` :smile: Kinda unfortunate for the ergonomics of using `t` but it is what it is :shrug: